dependent_types.md

```diff
@@ -0,0 +1,279 @@
+
+Third step in our wonderfull type land is 
+
+**Types depending on Types** (_type functions_)
+
+In order to demonstrate what we mean by this we will use Type Families. These [slides](https://cdepillabout.github.io/haskell-type-families-presentation/#/) do really good job in explaining what are type families.
+
+Basically they provide a way to define a function that operates on types. Depending on the concrete implementation we can return different types for different instances of the typeclass or at least we have that choice.
+
+This leads us to the final step to dependent types which is: 
+
+**Types depending on Values** (_dependent types_)
+
+As I mentioned at the beginning Haskell still does not have native support for dependent types but we have a handfull of extensions that provide a way to touch on that. We are now arriving to the next step in understanding dependent types which is
+
+## Local Assumptions
+
+GADTs or Generalized Algebraic DataTypes is the Haskell extension that provides us with _local assumptions_. What does that mean? 
+
+Well, let us look at some of the standard Haskell datatypes :
+
+```
+data Maybe a = Just a | Nothing
+data Either a b = Left a | Right b
+```
+If we rewrote these types with hypotetical syntax we would come to this:
+```
+data Maybe a = Just a
+     Maybe a = Nothing
+
+data Either a b = Left a
+     Either a b = Right b
+```
+and this can inform us that we are really limited with what we do on left side of the `=` sign. 
+
+Problem with this is that we don't get to do pattern matching on the type constructors which in turn means that we can only have free type variables on the left side.
+
+This is where *GADTs* language extension comes to play. It allows us to define type contraints on the type variable which limits the possible type variables that can be used to construct a type and later in the code have *local assumptions* when pattern matching on the data constructor which was used to constuct a type. 
+
+Lets take a look at some example code:
+
+```
+data IntOrString a where
+    IntConstructor    :: a ~ Int    => IntOrString a -- a has a type constraint to Int
+    StringConstructor :: a ~ String => IntOrString a -- a has a type constraint to String
+```
+This is desugared version of the more convenient syntax that we usually use
+
+```
+data IntOrString a where
+    IntConstructor    :: Int    -> IntOrString a
+    StringConstructor :: String -> IntOrString a
+```
+
+The classic `Maybe` datatype would look like this if written with GADTs
+
+```
+data Maybe a where
+    Just    :: a -> Maybe a
+    Nothing :: Maybe a
+```
+And here is example of the pattern match:
+
+```
+wasItStringOrInt :: IntOrString a -> String
+wasItStringOrInt x =
+    case x of
+        -- Local assumptions - Local because it is limited to only one case branch 
+        -- and assumption because we assume what type is our type parameter
+        IntConstructor a    -> "Constructed with Int"      -- a ~ Int
```

Comment:
I don't think this example is helpful because you don't utilize the local assumption in any way. I.e. you don't use the fact that now we know that `a ~ Int`. Perhaps do something like `show (a + 1)`, because it's clear that we can't `+1` a string. Same goes for the other case branch, perhaps try doing `reverse` there or something?

Reply:
Even then, the purpose of a GADT is not clear here. You could use `Either Int String` and write the same example :)

Reply:
Try using an example where the type variable is mentioned twice, i.e. `f :: IntOrString a -> IntOrString a -> ...`. Then you'd know that both values are of the same type, and you could apply a binary operation (let's say addition for `Int`, concatenation for `String`)